ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+    
+    Restritoin: restrictions for this function
+    """
+    # Note, innerVariable only exists inside this function
+    # if you call innerVariable outside of the function, error will show 
+    # That this variable does not exist
+    innerVariable = 'something'
+    # But the function can access variables outside of itself
+    x = 1 
+    return 'this is the output of the function'
+
+#when you call a function, you are typing its name and input the parameter
+FunctionName("input")
+
+# You can do math or print multiple things in a single line of print()
+x = 9.81
+print(3.53+5, "hello"+" world",x," m/s^2")
+
+"""
+While we are on the topic of function, lets talk about check 
+
+import check 
+
+check.expect("name of this test",function(parameter),expect Output)
+check.within("the name",function(parameter),expect output, error range)
+
+I hate it too much, that's all I want to say about it 
+"""
+
+"""
+Comment 
+
+#this thing is a comment
+"""
+
+#Comment will not be read by the computer
+v = "aer" #You can put comment anywhere 
+##Its awesome
+
+"""
+Algrebratical Expressions
+
++ - addition
+- - subtraction 
+* - multiplication
+/ - division 
+% - Modules
+** - Exponentiation
+// = Floor Division
+
+Remember that there are two types of numbers 
+int: Intergers, whole numbers 
+float: Decimal numbers 
+
+if you want more math, you have to use the math library 
+
+import math 
+
+then you can do all your wacky math.cos(),math.sine(),math.sqrt() stuff 
+
+Down there are the example of some 
+"""
+# Module is the remain after division
+# for example, 2/2 is 1, there is no remainder
+# 3/2 can also be see as 2+1/2, the remainder will be 1 
+print(2%2)
+print(3%2)
+
+# be aware of the difference in decimal places
+print(2//3)
+print(2/3)
+
+# you can also use + and * on string 
+word = "hello" + " world"
+hohoho = 'ho'*3
+
+"""
+Assignment Operators (in-place operators)
+You dont have to use them, but just know what they are 
+
+= -> x = 5 
++= -> x += 3 is the same sa x = x + 3 
+-= -> x -= 3 is the same as x = x - 3 
+*= -> x *= 3 is the same sa x = x * 3 
+/= -> x /= 3 is the same sa x = x / 3 
+%= -> x %= 3 is the same sa x = x % 3 
+**= -> x **= 3 is the same sa x = x ** 3 
+
+there are more than these, but just know the onces above should be enough 
+for you to go through this course
+
+"""
+
+# You cannot use two assignment operators on the same variable at the same line 
+# But you can mix it with other algebretical operators 
+x += 435+493/2302567
+
+"""
+Everything below will return a boolean value(True or False)
+
+Comparison Oprators 
+
+Operator   Name           Example
+==         Equal          x == y, 2 == 1 
+!=       Not Equal       x != 7, y != y
+>      Greater than       3 > 2, 4 > 5
+<       Less than         1 < 2, 5 < 5 
+>=   Greater or Equal to   1 >= 1, 1 >= 4
+<=   less or equal to      x <= 3 , 4 <= 4 
+
+You usually use them with if statement or while statement
```

Comment:
Another operation to consider is the `is` comparison operator. This checks if two variables point to the same location in memory. For example, if you have two strings:
```python
str1 = "string one"
str2 = "string one"
```
even though the contents of the strings are identical, these are two different objects. If you are dealing with multiple variables that can have the same value if you want to check if two variables point to the same object, use is. Here's an example where we create a new variable that has its value passed by reference (i.e. with pointers).
```python
a = 1
b = a

# The is operator will return True
print(a is b)
```

If you want to check if a variable is of a certain type, there is a built-in python function for doing so, called `isinstance`. So instead of writing:
```python
var = 1
# ... do stuff involving var 1

# use an if block that runs if var being an integer
if type(var) == int:
    # blah blah blah
```
you should use:
```python
var = 1
# ... do stuff involving var 1

# use an if block that runs if var is an integer
if isinstance(var, int):
    # blah blah blah
```